lib/Monads/trace/Trace_More_RG.thy

```diff
@@ -138,6 +138,24 @@ lemma assertE_tsp:
   "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> assertE Q \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. Q \<and> P s0 s\<rbrace>,\<lbrace>E\<rbrace>"
   by (wpsimp wp: assertE_wp)
 
+lemma gets_inv_rg[simp]:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> gets f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>_. P\<rbrace>"
+  by (simp add: gets_def, wp)
```

Comment:
Do we want to replicate the `simp`/`iff` etc attributes in the trace monad? Might be our chance for getting a cleaner setup on the trace side.

Reply:
Good point, although in that case I'd be tempted to remove these inv rules again. In these cases the standard `wp` rules should be fine, if not better.